ast/compile_test.go

```diff
@@ -4252,3 +4265,279 @@ func TestCompilerPassesTypeCheckNegative(t *testing.T) {
 		t.Fatal("Incorrectly detected a type-checking violation")
 	}
 }
+
+func testParseSchema(t *testing.T, schema, expectedType string) {
+	var sch interface{}
+	err := util.Unmarshal([]byte(schema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	newtype, err := setTypesWithSchema(sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if newtype == nil {
+		t.Fatalf("parseSchema returned nil type")
+	}
+	if newtype.String() != expectedType {
+		t.Fatalf("parseSchema returned an incorrect type: %s", newtype.String())
+	}
+}
+
+func TestParseSchemaObject(t *testing.T) {
+	testParseSchema(t, objectSchema, "object<b: array<object<a: number, b: array<number>, c: any>>, foo: string>")
+}
+
+func TestSetTypesWithSchemaRef(t *testing.T) {
+	var sch interface{}
+	err := util.Unmarshal([]byte(podSchema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	newtype, err := setTypesWithSchema(sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if newtype == nil {
+		t.Fatalf("parseSchema returned nil type")
+	}
+	if newtype.String() == "object<apiVersion: string, kind: string, metadata: any, spec: any, status: any>" {
+		t.Fatalf("parseSchema returned an incorrect type: %s", newtype.String())
+	}
+
+}
+
+func TestParseSchemaUntypedField(t *testing.T) {
+	testParseSchema(t, untypedFieldObjectSchema, "object<foo: any>")
+}
+
+func TestParseSchemaNoChildren(t *testing.T) {
+	testParseSchema(t, noChildrenObjectSchema, "object[any: any]")
+}
+
+func TestParseSchemaArrayNoItems(t *testing.T) {
+	testParseSchema(t, arrayNoItemsSchema, "object<b: array[any]>")
+}
+
+func TestParseSchemaBooleanField(t *testing.T) {
+	testParseSchema(t, booleanSchema, "object<a: boolean>")
+}
+
+func TestCompileSchemaEmptySchema(t *testing.T) {
+	schema := ""
+	var sch interface{}
+	err := util.Unmarshal([]byte(schema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	jsonSchema, _ := compileSchema(sch)
+	if jsonSchema != nil {
+		t.Fatalf("Incorrect return from parseSchema with an empty schema")
+	}
+}
+
+func TestParseSchemaWithSchemaBadSchema(t *testing.T) {
+	var sch interface{}
+	err := util.Unmarshal([]byte(objectSchema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	jsonSchema, err := compileSchema(sch)
+	if err != nil {
+		t.Fatalf("Unable to compile schema")
+	}
+	newtype, err := parseSchema(jsonSchema) // Did not pass the subschema
+	if newtype != nil {
+		t.Fatalf("Incorrect return from parseSchema with a bad schema")
+	}
+}
+
+func TestWithSchema(t *testing.T) {
+	c := NewCompiler().
+		WithCapabilities(&Capabilities{Builtins: []*Builtin{Split}})
```

Comment:
I don't yet understand the significance of having this builtin here...?

Reply:
Not needed indeed. Will remove, thanks.

Reply:
Done  (commit https://github.com/open-policy-agent/opa/pull/3060/commits/bd984a0fcc39b763449b8ea3616d419eeb925ab5)